main.cpp

```diff
@@ -38,25 +40,330 @@ write 3 UDTs below (where it asks for Copied UDTs) that EACH have:
+GraphicalRepresentation::GraphicalRepresentation(Color newBackgroundColor)
+{
+    backgroundColor = newBackgroundColor;
+}
+GraphicalRepresentation::~GraphicalRepresentation()
+{
+    cout << "GraphicalRepresentation-Dtor" << endl;
+}
+
+bool GraphicalRepresentation::moveToXCoord(int newXCoord, int stepsize)
+{
+    while (xCoordinate < newXCoord)
+    {
+        xCoordinate += stepsize;
+        cout << "xCoordinate = " << xCoordinate << endl;
+    }
+    return true;
+}
+bool GraphicalRepresentation::moveToYCoord(int newYCoord, int stepsize)
+{
+    while (yCoordinate < newYCoord){
+        yCoordinate += stepsize;
+        cout << "yCoordinate = " << yCoordinate << endl;
+    }
+    return true;
+}
+void GraphicalRepresentation::attachLeftOf(GraphicalRepresentation targetRepresentation)
+{
+    xCoordinate = targetRepresentation.xCoordinate - width;
+    yCoordinate = targetRepresentation.yCoordinate;
+    cout << "New xCoordinate = " << xCoordinate << ", new yCoordinate = " << yCoordinate << endl;
+}
+
 
 /*
  UDT 2:
  */
+struct Oscillator
+{
+    struct OscillatorType
+    {
+        enum Waveform {sine, triangle, pulse, whiteNoise, pinkNoise};
+
+        Waveform mWaveform;
+        float dutyCycle = 0.0f;
+        int nrOfVoices = 1;
+        float detune = 0.0f;
+        int phase = 0;
+
+        OscillatorType()
+        {
+            mWaveform = pulse;
+            dutyCycle = 0.5f;
+        }
+        OscillatorType(Waveform pWaveform) : mWaveform(pWaveform) {}
+        ~OscillatorType() { cout << "OscillatorType-Dtor" << endl; }
+
+        void cycleThroughNrOfVoices();
+        int setNewPhase(int newPhase);
+        bool changeWaveform(Waveform pWaveform);
+    };
+
+    OscillatorType mOscillatorType;
+    int octave = 0;
+    int semitones = 0;
+    float coarse = 0;
+    float fine = 0;
+    float samplingFreq = 44100;
+    GraphicalRepresentation oscRepresentation { GraphicalRepresentation::grey };
+
+    Oscillator();
+    Oscillator(OscillatorType pOscillatorType);
+    ~Oscillator();
+    
+    void playSound(bool MIDISignal);
+    int convertOctavesToSemitones(int octaves);
+    float setSamplingFrequency(float newSamplingFrequency);
+};
+
+
+void Oscillator::OscillatorType::cycleThroughNrOfVoices()
+{
+    for (int i = 1; i <= 16; ++i)
+    {
+        cout << "Number of voices = " << i << endl;
+    }
+}
+int Oscillator::OscillatorType::setNewPhase(int newPhase)
+{
+    phase = newPhase;
+    return phase;
+}
+bool Oscillator::OscillatorType::changeWaveform(Waveform pWaveform){
+    cout << "Changing Waveform..." << endl;
+    mWaveform = pWaveform;
+    return true;
+}
+
+Oscillator::Oscillator() : mOscillatorType(OscillatorType()) {}
+Oscillator::Oscillator(OscillatorType pOscillatorType)
+{
+    mOscillatorType = OscillatorType(pOscillatorType);
+}
```

Comment:
This is making an `OscillatorType` from the function parameter, which is already an `OscillatorType`, and then assigning it to your member variable.
that's redundant.  You're making copies of copies.
- [ ] revise